### DISCUSSION TITLE
Deleting data with @@unique condition, null is not allowed?

### DISCUSSION THREAD

**Discussion by hopoduck** [prisma/prisma#23522] [Q&A] [answered]
### Question

The prisma service is used in the nest.js project.

I deleted one reaction data from Prisma using a unique condition, but I got an error saying commentId cannot be null. But actually commentId is a nullable column.

What should I do?

### Expected behavior (optional)

Delete successful

### Information about Prisma Schema, Client Queries and Environment (optional)

- schema.prisma
```prisma
model Reaction {
  id        String   @id @default(cuid())
  articleId String?
  commentId String?
  name      String
  userToken String
  article   Article? @relation(fields: [articleId], references: [id])
  comment   Comment? @relation(fields: [commentId], references: [id])
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  @@unique([articleId, commentId, name, userToken])
}
```

- Prisma Client queries 
```ts
    return await this.prisma.reaction.delete({
      where: {
        articleId_commentId_name_userToken: deleteReactionDto,
      },
    });
  }
```

- Error log
```
  59 async deleteReaction(deleteReactionDto: DeleteReactionDto) {
  60   this.logger.debug(`deleteReaction >> ${JSON.stringify(deleteReactionDto)}`);
→ 61   return await this.prisma.reaction.delete({
         where: {
           articleId_commentId_name_userToken: {
             name: "fluent-emoji:face-screaming-in-fear",
             articleId: "cltiw4xca00008piafjo87tvv",
             userToken: "b463c7ebe8df41158b28d7cee56f1d6d",
       +     commentId: String
           }
         }
       })

Argument `commentId` must not be null.
```

- Current data
![image](https://github.com/prisma/prisma/assets/64189819/895dc11d-8618-45be-add1-fb39635d0b76)

Actually, `articleId` / `commentId` are nullable values. How can I delete that data?

Environment variables loaded from .env
prisma                  : 5.11.0
@prisma/client          : 5.11.0
Computed binaryTarget   : linux-arm64-openssl-3.0.x
Operating System        : linux
Architecture            : arm64
Node.js                 : v18.19.0
Query Engine (Node-API) : libquery-engine efd2449663b3d73d637ea1fd226bafbcf45b3102 (at node_modules/.pnpm/@prisma+engines@5.11.0/node_modules/@prisma/engines/libquery_engine-linux-arm64-openssl-3.0.x.so.node)
Schema Engine           : schema-engine-cli efd2449663b3d73d637ea1fd226bafbcf45b3102 (at node_modules/.pnpm/@prisma+engines@5.11.0/node_modules/@prisma/engines/schema-engine-linux-arm64-openssl-3.0.x)
Schema Wasm             : @prisma/prisma-schema-wasm 5.11.0-15.efd2449663b3d73d637ea1fd226bafbcf45b3102
Default Engines Hash    : efd2449663b3d73d637ea1fd226bafbcf45b3102
Studio                  : 0.499.0

**Comment by ludralph** [comment, marked as answer]
Hi @hopoduck 👋 

Thank you for raising this question.

Prisma is expecting a `commentId` value in the `where` clause of your `delete` query, even though `commentId` is a nullable field in your schema. This is because you're using a composite unique constraint `(@@unique([articleId, commentId, name, userToken]))` in your `Reaction` model. When you're deleting a record using this composite unique identifier, Prisma expects all fields in the unique constraint to be provided.

If you want to delete a `Reaction` based on a subset of these fields, you might need to use a different approach. For example, you could first perform a `findMany` operation with the fields you have, and then delete each returned record individually.
For example:
```js
const reactions = await this.prisma.reaction.findMany({
  where: {
    articleId: deleteReactionDto.articleId,
    name: deleteReactionDto.name,
    userToken: deleteReactionDto.userToken,
  },
});

for (const reaction of reactions) {
  await this.prisma.reaction.delete({
    where: { id: reaction.id },
  });
}
```

In this example, we first find all `Reaction` records that match the `articleId`, `name`, and `userToken` fields from your `deleteReactionDto`. We then delete each of these records individually based on their `id`.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by hopoduck** [reply]
Hello!

I was wondering if there was another way to use this feature. 
I used all `@@unique` conditions to run deleteMany and set it to delete only one data.

Thank you for your assistance.
